File: Python/0572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Time O(m*n), m is the number of nodes of s, n is the number of nodes of t
# Space O(h), h is the height of s
class Solution:
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:

        def isSame(x, y):
            if not x and not y:
                return True

            if not x or not y:
                return False

            return (x.val == y.val) and isSame(x.left, y.left) and isSame(x.right, y.right)

        def preOrderTraverse(s, t):
            return s and (isSame(s, t) or preOrderTraverse(s.left, t) or preOrderTraverse(s.right, t))

        return preOrderTraverse(s,t)



class Solution:
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        def isSame(s,t):
            if not s and not t:
                return True
            if not s or not t:
                return False
            return s.val == t.val and isSame(s.left, t.left) and isSame(s.right, t.right)

        return s and (isSame(s, t) or self.isSubtree(s.left, t) or self.isSubtree(s.right, t))

"""
We are checking if a tree rooted at s is same as tree rooted at t. If not,
we check if tree rooted at s.left or if tree rooted at s.right is same as t.
We keep doing this until we have checked every node in tree s.
"""

File: C++/0572-subtree-of-another-tree.cpp
// Tags: Tree Amazon
// Time: O(m*n), m is the number of nodes of s, n is the number of nodes of t
// Space: O(h), h is the height of s
class Solution {
public:
    bool isSubtree(TreeNode* s, TreeNode* t) {
        return preOrderTraverse(s, t);

    }

private:
    bool isSame(TreeNode *x, TreeNode *y){
        if (!x && !y)
            return true;

        if (!x || !y)
            return false;

        return x->val == y->val && isSame(x->left, y->left) && isSame(x->right, y->right);
    }

    bool preOrderTraverse(TreeNode *s, TreeNode *t) {
        return s && (isSame(s, t) || preOrderTraverse(s->left, t) || preOrderTraverse(s->right, t));
    }
};




class Solution {
    bool isSame(TreeNode *s, TreeNode *t) {
        if (!s && !t)
            return true;
        if (!s || !t)
            return false;
        return s->val == t->val && isSame(s->left, t->left) && isSame(s->right, t->right);
    }

public:
    bool isSubtree(TreeNode* s, TreeNode* t) {
        return s && (isSame(s, t) || isSubtree(s->left, t) || isSubtree(s->right, t));
    }
};

